x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java

```diff
@@ -0,0 +1,451 @@
+                    return 0;
+                }
+
+                // TODO could we "just" get the memory of the array and dump it?
+                BytesRef scratch = new BytesRef();
+                long length = 0;
+                for (int i = 0; i < block.getPositionCount(); i++) {
+                    BytesRef v = block.getBytesRef(i, scratch);
+
+                    out.write(v.bytes, v.offset, v.length);
+                    length += v.length;
+                }
+                return length;
+            });
+        }
+
+        private static int offsetVectorLength(BytesRefBlock block) {
+            return Integer.BYTES * (block.getPositionCount() + 1);
+        }
+
+        private int dataVectorLength(BytesRefBlock block) {
+            if (block.areAllValuesNull()) {
+                return 0;
+            }
+
+            // TODO we can probably get the length from the vector without all this sum
+
+            int length = 0;
+            BytesRef scratch = new BytesRef();
+            for (int i = 0; i < block.getPositionCount(); i++) {
+                BytesRef v = block.getBytesRef(i, scratch);
+                length += v.length;
+            }
+            return length;
+        }
+    }
+
+    /**
+     * Conversion of ByteRefs where each value is itself converted to a different format.
+     */
+    public static class TransformedBytesRef extends BytesRefConverter {
+
+        private final BiFunction<BytesRef, BytesRef, BytesRef> valueConverter;
+
+        /**
+         *
+         * @param esqlType ESQL type name
+         * @param minorType Arrow type
+         * @param valueConverter a function that takes (value, scratch) input parameters and returns the transformed value
+         */
+        public TransformedBytesRef(String esqlType, Types.MinorType minorType, BiFunction<BytesRef, BytesRef, BytesRef> valueConverter) {
+            super(esqlType, minorType);
+            this.valueConverter = valueConverter;
+        }
+
+        @Override
+        public void convert(Block b, List<ArrowBuf> bufs, List<BufWriter> bufWriters) {
+            BytesRefBlock block = (BytesRefBlock) b;
+            BytesRefBlock transformed = transformValues(block);
+            super.convert(transformed, bufs, bufWriters);
+        }
+
+        /**
+         * Creates a new BytesRefBlock by applying the value converter to each non null and non empty value
+         */
+        private BytesRefBlock transformValues(BytesRefBlock block) {
+            try (BytesRefBlock.Builder builder = block.blockFactory().newBytesRefBlockBuilder(block.getPositionCount())) {
+                BytesRef scratch = new BytesRef();
+                for (int i = 0; i < block.getPositionCount(); i++) {
+                    if (block.isNull(i)) {
+                        builder.appendNull();
+                    } else {
+                        BytesRef bytes = block.getBytesRef(i, scratch);
+                        if (bytes.length != 0) {
```

Comment:
Now that I think about it, I think this should apply to empty `BytesRef` values. Let the converter decide what to do.